addons/website_theme_install/models/ir_module_module.py

```diff
@@ -203,43 +227,85 @@ def _unload_one_theme_module(self, website):
+        return all_mods
+
+    def _get_stream_website_ids(self):
+        """
+            Websites for which this theme (self) is in the stream of their theme.
+        """
+        websites = self.env['website'].search([])
+        for website in websites:
+            for theme in website.theme_id._get_stream_themes():
+                if self == theme:
+                    if website not in websites:
+                        websites += website
+        return websites
+
+    def _get_downstream_website_ids(self):
+        """
+            Websites for which this theme (self) is in the downstream of their theme.
+        """
         self.ensure_one()
-        installed_deps = self + website.theme_id.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
-        for mod in installed_deps:
-            mod._unload_one_theme_module(website)
+        websites = self.env['website'].search([])
+        for website in websites:
+            for theme in website.theme_id._get_downstream_themes():
+                if self == theme:
+                    if website not in websites:
+                        websites += website
+        return websites
 
-    @api.multi
-    def _add_theme_on_website(self, website):
-        self.ensure_one()
-        mods_to_load = reversed(self + self.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_')))
-        for mod in mods_to_load:
-            _logger.info('Load theme %s for website %s from template.' % (mod.name, website.id))
-            mod._load_one_theme_module(website, with_update=False)
-            self.env['theme.utils']._post_copy(mod)
+    def _remove_theme_on_website(self, website):
+        for mod in reversed(website.theme_id._get_stream_themes()):
+            mod._unload_one_theme_module(website)
+        website.theme_id = False
 
     @api.multi
     def button_choose_theme(self):
         self.ensure_one()
         website = self.env['website'].get_current_website()
 
-        # Unload previous theme
-        if website.theme_id:
-            website.theme_id._remove_theme_on_website(website)
-
-        website.theme_id = self
+        self._remove_theme_on_website(website)
 
         # create template data of missing theme
         next_action = False
         if self.state != 'installed':
+            # upgrade the upstream chain first
+            themes = self._get_upstream_themes()
+            if themes[0].state == 'installed':
+                themes[0].button_immediate_upgrade()
+
             next_action = self.button_immediate_install()
             # reload registry to check if 'theme.utils'._<theme_name>_post_copy exists e.g.
             self.env.reset()
             self = self.env()[self._name].browse(self.id)
 
+        website.theme_id = self
         # Copy new theme from template table to real table
-        self._add_theme_on_website(website)
+        for mod in self._get_stream_themes():
```

Comment:
in the current state or the PR, this is only needed if self was already installed. Otherwise the upgrade/install of the chain will trigger if (if we move the website.theme_id = self above the install).